scipy/optimize/_highs/pyHiGHS/src/highs_wrapper.pyx

```diff
@@ -0,0 +1,621 @@
+    HighsLp,
+    HighsSolution,
+    HighsModelStatus,
+    HighsModelStatusNOTSET,
+    HighsModelStatusLOAD_ERROR,
+    HighsModelStatusMODEL_ERROR,
+    HighsModelStatusMODEL_EMPTY,
+    HighsModelStatusPRESOLVE_ERROR,
+    HighsModelStatusSOLVE_ERROR,
+    HighsModelStatusPOSTSOLVE_ERROR,
+    HighsModelStatusPRIMAL_INFEASIBLE,
+    HighsModelStatusPRIMAL_UNBOUNDED,
+    HighsModelStatusOPTIMAL,
+    HighsModelStatusREACHED_DUAL_OBJECTIVE_VALUE_UPPER_BOUND,
+    HighsModelStatusREACHED_TIME_LIMIT,
+    HighsModelStatusREACHED_ITERATION_LIMIT,
+)
+from HighsInfo cimport HighsInfo
+from HighsOptions cimport (
+    HighsOptions,
+    OptionRecord,
+    OptionRecordBool,
+    OptionRecordInt,
+    OptionRecordDouble,
+    OptionRecordString,
+)
+
+# options to reference for default values and bounds;
+# make a map to quickly lookup
+cdef HighsOptions _ref_opts
+cdef cppmap[string, OptionRecord*] _ref_opt_lookup
+cdef OptionRecord * _r = NULL
+for _r in _ref_opts.records:
+    _ref_opt_lookup[_r.name] = _r
+
+cdef str _opt_warning(string name, val, valid_set=None):
+    cdef OptionRecord * r = _ref_opt_lookup[name]
+
+    # BOOL
+    if r.type == HighsOptionTypeBOOL:
+        default_value = (<OptionRecordBool*> r).default_value
+        return ('Option "%s" is "%s", but only True or False is allowed. '
+                'Using default: %s.' % (name.decode(), str(val), default_value))
+
+    # INT
+    if r.type == HighsOptionTypeINT:
+        lower_bound = int((<OptionRecordInt*> r).lower_bound)
+        upper_bound = int((<OptionRecordInt*> r).upper_bound)
+        default_value = int((<OptionRecordInt*> r).default_value)
+        if upper_bound - lower_bound < 10:
+            int_range = str(set(range(lower_bound, upper_bound + 1)))
+        else:
+            int_range = '[%d, %d]' % (lower_bound, upper_bound)
+        return ('Option "%s" is "%s", but only values in %s are allowed. '
+                'Using default: %d.' % (name.decode(), str(val), int_range, default_value))
+
+    # DOUBLE
+    if r.type == HighsOptionTypeDOUBLE:
+        lower_bound = (<OptionRecordDouble*> r).lower_bound
+        upper_bound = (<OptionRecordDouble*> r).upper_bound
+        default_value = (<OptionRecordDouble*> r).default_value
+        return ('Option "%s" is "%s", but only values in (%g, %g) are allowed. '
+                'Using default: %g.' % (name.decode(), str(val), lower_bound, upper_bound, default_value))
+
+    # STRING
+    if r.type == HighsOptionTypeSTRING:
+        if valid_set is not None:
+            descr = 'but only values in %s are allowed. ' % str(set(valid_set))
+        else:
+            descr = 'but this is an invalid value. %s. ' % r.description.decode()
+        default_value = (<OptionRecordString*> r).default_value.decode()
+        return ('Option "%s" is "%s", '
+                '%s'
+                'Using default: %s.' % (name.decode(), str(val), descr, default_value))
+
+    # We don't know what type?
```

Comment:
Do you have tests that reach into this Cython code? IIRC it can be annoying to do line coverage analysis for Cython, esp. in a project as complex as SciPy.

Reply:
The final return statement in this function should be unreachable, it is added here as a sanity check and protection against any future breakage due to upstream changes in type handling.  The warnings are tested [here](https://github.com/mdhaber/scipy/blob/linprog_highs2/scipy/optimize/tests/test_linprog.py#L1482)